### PR TITLE
mgr/dashboard: Compare values of MTU alert by device

### PR DESCRIPTION
**Pull request by p-se**
Fixes: https://tracker.ceph.com/issues/55004

Signed-off-by: Patrick Seidensal <pseidensal@suse.com>

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [x] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: monitoring/ceph-mixin/prometheus_alerts.yml [hunk 1]
```diff
@@ -704,15 +704,26 @@ groups:
...
+            scalar(
+              max by (device) (node_network_mtu_bytes) !=
+                quantile by (device) (.5, node_network_mtu_bytes)
+            )
+          or
+          node_network_mtu_bytes * (node_network_up{device!="lo"} > 0) ==
+            scalar(
+              min by (device) (node_network_mtu_bytes) !=
+                quantile by (device) (.5, node_network_mtu_bytes)
+            )
```
Comment (aaSharma14) [inline monitoring/ceph-mixin/prometheus_alerts.yml:718]:
```suggestion
        expr: |
          node_network_mtu_bytes{device!="lo"} * (node_network_up{device!="lo"} > 0) ==
            scalar(
              max by (device) (node_network_mtu_bytes{device!="lo"} * (node_network_up{device!="lo"} > 0)) !=
                quantile by (device) (.5, node_network_mtu_bytes{device!="lo"} * (node_network_up{device!="lo"} > 0))
            )
          or
          node_network_mtu_bytes{device!="lo"} * (node_network_up{device!="lo"} > 0) ==
            scalar(
              min by (device) (node_network_mtu_bytes{device!="lo"} * (node_network_up{device!="lo"} > 0)) !=
                quantile by (device) (.5, node_network_mtu_bytes{device!="lo"} * (node_network_up{device!="lo"} > 0))
            )
```

Comment (aaSharma14) [inline monitoring/ceph-mixin/prometheus_alerts.yml:718]:
```suggestion
        expr: |
          node_network_mtu_bytes * (node_network_up{device!="lo"} > 0) ==
            scalar(
              max by (device) (node_network_mtu_bytes * (node_network_up{device!="lo"} > 0)) !=
                quantile by (device) (.5, node_network_mtu_bytes * (node_network_up{device!="lo"} > 0))
            )
          or
          node_network_mtu_bytes * (node_network_up{device!="lo"} > 0) ==
            scalar(
              min by (device) (node_network_mtu_bytes * (node_network_up{device!="lo"} > 0)) !=
                quantile by (device) (.5, node_network_mtu_bytes * (node_network_up{device!="lo"} > 0))
            )
```

Comment (aaSharma14) [reply monitoring/ceph-mixin/prometheus_alerts.yml:718]: Thanks @p-se for this PR. Am in process of writing the promtool test for this, I think we need to consider only the devices that are in up state in calculations as well as I found some use cases failing becaiuse of this.

**Comment by aaSharma14** [comment]
jenkins test dashboard

**Comment by aaSharma14** [comment]
jenkins test dashboard

**Review by nizamial09 — APPROVED** [review]

**Comment by p-se** [comment]
Does this PR need to go through Teuthology?

File: monitoring/ceph-mixin/prometheus_alerts.yml [hunk 2]
```diff
@@ -704,15 +704,26 @@ groups:
...
+            scalar(
+              max by (device) (node_network_mtu_bytes * (node_network_up{device!="lo"} > 0)) !=
+                quantile by (device) (.5, node_network_mtu_bytes * (node_network_up{device!="lo"} > 0))
+            )
+          or
+          node_network_mtu_bytes * (node_network_up{device!="lo"} > 0) ==
+            scalar(
+              min by (device) (node_network_mtu_bytes * (node_network_up{device!="lo"} > 0)) !=
+                quantile by (device) (.5, node_network_mtu_bytes * (node_network_up{device!="lo"} > 0))
+            )
```
Comment (epuertat) [inline monitoring/ceph-mixin/prometheus_alerts.yml:718]: 🤯  this is well beyond my PromQL literacy 🙈 well, I'll trust the unit tests, which seem to make sense!

**Review by epuertat — APPROVED** [review]
LGTM! Thanks a lot @p-se !

We discussed at today's stand-up whether to improve the MTU alert in a follow-up PR and we agreed on:
* as per this PR the MTU alert will work as long as all hosts are using the same MTU setting for the same interface name across the cluster (i.e.: `eth0` in all hosts has MTU 1500, `eth1` has MTU 9000, etc).
* to make this work with the network address instead of with the interface name, we'd need to provide new metrics via mgr/prometheus (since node-exporter doesn't provide this info yet).
* based on upstream users and downstream customer experience so far, we think that it is fair to assume that Ceph users are generally configuring the same networks (cluster or public) on network interfaces whose names are the same.
* for those users don't following that rule, the impact of this alert would be minimal, as it could be easily muted from the Dashboard as soon as it would be first triggered misleadingly.

**Pull request merged**